packages/babel-plugin-transform-typescript/src/index.js

```diff
@@ -0,0 +1,266 @@
+function isInType(path) {
+  switch (path.parent.type) {
+    case "TSTypeReference":
+    case "TSQualifiedName":
+    case "TSExpressionWithTypeArguments":
+    case "TSTypeQuery":
+      return true;
+    default:
+      return false;
+  }
+}
+
+function isImportTypeOnly(binding, sourceFileHasJsx) {
+  for (const path of binding.referencePaths) {
+    if (!isInType(path)) {
+      return false;
+    }
+  }
+
+  if (binding.identifier.name != "React") {
+    return true;
+  }
+
+  return !sourceFileHasJsx;
+}
+
+interface State {
+  sourceFileHasJsx: boolean,
+}
+
+export default function({ types: t }) {
+  return {
+    inherits: syntaxTypeScript,
+    visitor: {
+      //"Pattern" alias doesn't include Identifier or RestElement.
+      Pattern: visitPattern,
+      Identifier: visitPattern,
+      RestElement: visitPattern,
+
+      Program(path, state: State) {
+        state.sourceFileHasJsx = false;
+        path.traverse({
+          JSXOpeningElement() {
+            state.sourceFileHasJsx = true;
+          },
+          JSXElement() {
+            state.sourceFileHasJsx = true;
+          },
+        });
+      },
+
+      ImportDeclaration(path, state: State) {
+        // Note: this will allow both `import { } from "m"` and `import "m";`.
+        // In TypeScript, the former would be elided.
+        if (path.node.specifiers.length === 0) {
+          return;
+        }
+
+        let allElided = true;
+        const importsToRemove: Path<Node>[] = [];
+
+        for (const specifier of path.node.specifiers) {
+          const binding = path.scope.getBinding(specifier.local.name);
+          if (isImportTypeOnly(binding, state.sourceFileHasJsx)) {
+            importsToRemove.push(binding.path);
+          } else {
+            allElided = false;
+          }
+        }
+
+        if (allElided) {
+          path.remove();
+        } else {
+          for (const importPath of importsToRemove) {
+            importPath.remove();
+          }
+        }
+      },
+
+      TSDeclareFunction(path) {
+        path.remove();
+      },
+
+      TSDeclareMethod(path) {
+        path.remove();
+      },
+
+      VariableDeclaration(path) {
+        if (path.node.declare) path.remove();
+      },
+
+      ClassMethod(path) {
+        const { node } = path;
+
+        if (node.accessibility) node.accessibility = null;
+        if (node.abstract) node.abstract = null;
+        if (node.optional) node.optional = null;
+
+        if (node.kind !== "constructor") {
+          return;
+        }
+
+        // Collect parameter properties
+        const parameterProperties = [];
+        for (const param of node.params) {
+          if (param.type === "TSParameterProperty") {
+            parameterProperties.push(param.parameter);
+          }
+        }
+
+        if (!parameterProperties.length) {
+          return;
+        }
+
+        const assigns = parameterProperties.map(p => {
+          let name;
+          if (p.type === "Identifier") {
```

Comment:
`t.isIdentifier(p)`. Doing explicit comparisons against `node.type` is an anti-pattern.